pytensor/tensor/subtensor.py

```diff
@@ -162,13 +187,21 @@ def as_index_literal(
     ------
     NotScalarConstantError
     """
-    if idx == np.newaxis or isinstance(getattr(idx, "type", None), NoneTypeT):
-        return np.newaxis
+    if isinstance(idx, int | np.integer):
+        return idx
+
+    # NOTE: np.newaxis is None
+    idxtype = getattr(idx, "type", None)
+    if idx is None or isinstance(idxtype, NoneTypeT):
+        return None
 
     if isinstance(idx, Constant):
-        return idx.data.item() if isinstance(idx, np.ndarray) else idx.data
+        data = idx.data
+        if isinstance(data, np.ndarray):
+            return cast(int, data.item())
+        return cast(int | None, data)
```

Comment:
```suggestion
        return cast(int | slice, data)
```
I think this is the right cast

Reply:
Alternatively, if you move this check below the one next `if` you can be sure that these constant must be integer variables (you already ruled out None above, and then `SliceType` constants with the `if` below

Reply:
no longer applies